fix: Subtitle Positioning Issue During Menu Display

bk-bf opened the pull request:
## 📋 Summary

Fixed subtitle positioning bug where subtitles jumped to screen center when video player menu was displayed. Now uses dynamic menu height measurement for precise subtitle positioning just above the menu.

## 🐛 Problem
Subtitles rendered in the middle of the screen when the video player menu overlay was active, obscuring video content and making it annoying to watch.

## 🔧 Solution

### Key Changes
- **Dynamic height measurement**: Real-time calculation of actual menu dimensions
- **Intelligent positioning**: Places subtitles just above menu with collision detection
- **Smart fallback**: Uses 15% screen height when measurement unavailable
- **Respects user preferences**: Only adjusts subtitles that would overlap with menu

### Modified Files
- `lib_mpv.dart` - Dynamic subtitle positioning logic
- `video_player_controls.dart` - Menu height measurement
- `base_player.dart` - Updated interface for menuHeight parameter
- `lib_mdk.dart` - API compatibility updates

## ✅ Benefits
- **Pixel-perfect positioning** using actual menu dimensions
- **Minimal visual disruption** with subtitles positioned just above menu
- **Maintains user preferences** for high-positioned subtitles
- **Cross-platform consistency** with graceful fallback behavior

## 🧪 Testing
Verified on macOS with various scenarios including dynamic height calculation, menu toggling, high-positioned subtitles, fallback behavior, and edge cases.

## 🔄 Backwards Compatibility
✅ Fully compatible - no breaking changes, preserves existing subtitle behavior and user preferences.

**Resolves:** https://github.com/DonutWare/Fladder/issues/403

bk-bf commented:
@sourcery-ai review

bk-bf commented:
sourcery-ai review

SourceryAI commented on lib/wrappers/players/lib_mpv.dart:
```diff
@@ -224,24 +237,54 @@ class _VideoSubtitlesState extends ConsumerState<_VideoSubtitles> {
     super.dispose();
   }
 
```
**suggestion:** Subtitle offset calculation now considers menu height, but may not handle rapid menu size changes.

If the menu height can change dynamically, ensure the widget rebuilds and recalculates the subtitle offset as needed. Trigger a rebuild on menu height changes if this isn't already handled.

Suggested implementation:

```
    final settings = ref.watch(subtitleSettingsProvider);
    final padding = MediaQuery.of(context).padding;
    final menuHeight = ref.watch(menuHeightProvider);

```

```
    // Process subtitle text
    final text = subtitle.where((line) => line.trim().isNotEmpty).map((line) => line.trim()).join('\n');

    // Use menuHeight in subtitle offset calculation as needed
    // For example, if you pass offset to SubtitleText:
    // final subtitleOffset = calculateSubtitleOffset(menuHeight, ...);

```

- You must ensure that `menuHeightProvider` exists and emits the current menu height. If not, create it and update it whenever the menu height changes.
- Update the subtitle offset calculation and pass the offset to the subtitle rendering widget (e.g., `SubtitleText`) as needed.
- If you use a different state management solution for menu height, adapt the `ref.watch(menuHeightProvider)` line accordingly.

SourceryAI reviewed (commented):
Hey @bk-bf - I've reviewed your changes - here's some feedback:

- Consider scheduling a post-frame callback (e.g. WidgetsBinding.addPostFrameCallback) to capture and store the menu height in state so subtitles update immediately once the bottom controls are laid out, preventing stale or null values on initial render.
- The nested math.max(math.min(...)) clamping in _calculateSubtitleOffset can be simplified to a single clamp operation to make the logic more readable.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:
## Overall Comments
- Consider scheduling a post-frame callback (e.g. WidgetsBinding.addPostFrameCallback) to capture and store the menu height in state so subtitles update immediately once the bottom controls are laid out, preventing stale or null values on initial render.
- The nested math.max(math.min(...)) clamping in _calculateSubtitleOffset can be simplified to a single clamp operation to make the logic more readable.

## Individual Comments

### Comment 1
<location> `lib/wrappers/players/lib_mpv.dart:239` </location>
<code_context>
-    if (widget.controller.player.platform?.configuration.libass ?? false) {
-      return const IgnorePointer(child: SizedBox.shrink());
-    } else {
-      return SubtitleText(
-        subModel: settings,
-        padding: padding,
</code_context>

<issue_to_address>
Subtitle offset calculation now considers menu height, but may not handle rapid menu size changes.

If the menu height can change dynamically, ensure the widget rebuilds and recalculates the subtitle offset as needed. Trigger a rebuild on menu height changes if this isn't already handled.

Suggested implementation:

```
    final settings = ref.watch(subtitleSettingsProvider);
    final padding = MediaQuery.of(context).padding;
    final menuHeight = ref.watch(menuHeightProvider);

```

```
    // Process subtitle text
    final text = subtitle.where((line) => line.trim().isNotEmpty).map((line) => line.trim()).join('\n');

    // Use menuHeight in subtitle offset calculation as needed
    // For example, if you pass offset to SubtitleText:
    // final subtitleOffset = calculateSubtitleOffset(menuHeight, ...);

```

- You must ensure that `menuHeightProvider` exists and emits the current menu height. If not, create it and update it whenever the menu height changes.
- Update the subtitle offset calculation and pass the offset to the subtitle rendering widget (e.g., `SubtitleText`) as needed.
- If you use a different state management solution for menu height, adapt the `ref.watch(menuHeightProvider)` line accordingly.
</issue_to_address>
~~~

</details>

***

Hi @bk-bf! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

PartyDonut reviewed (changes requested):
Hi thanks for taking a stab at implementing this.

Also make sure to check if it works on different displays, mobile for example is broken in both portrait and landscape.

<img height="800" alt="Image" src="https://github.com/user-attachments/assets/945bfccd-fa7a-4263-9c91-21a896cfdd55" />

Subtitles are overlayed by some UI elements


Also on a unrelated note, the initial comment seems to be at least partially generated by AI?
While it's always good to give as much information as needed I would personally prefer a more to the point note about the implemented changes.

bk-bf commented:
Hey @PartyDonut , 

I have implemented ur requested changes regarding the comment verbosity and removed the unnecessary test file. As for the height calculation, there is no easy way of migrating it from the video_player_controls.dart file into the _VideoSubtitles class inside of lib_mpv, without drastically changing my implementation logic and it seems more sensible to leave it in there for the following reasons: 

Why measurement should stay in video_player_controls.dart:

- Ownership - Controls widget owns and renders the menu UI
- Access - Only this widget has direct access to the menu's RenderBox
- Separation of concerns - Subtitle widgets shouldn't know about control UI structure
- Reusability - Different players get the same accurate measurement
- Clean architecture - Controls handle UI measurement, players handle positioning

Why video players should receive the measurement:

- Accurate positioning - Subtitle widgets get pre-measured data instead of guessing
- Fallback safety - When null, the calculator uses safe fallback values
- Player independence - Each player type can handle subtitles without measuring external UI

but maybe i misunderstood ur point, could u give me a clarification on that? Regarding the subtitle positioning issue on mobile, especially in vertical mode, to counteract that i have set the padding to 0 (previously at -0.03 to place the subs closer to the player menu bar), but left it for a possible future implementation of a user-adjustable slider to control subtitle positioning relative to the player menu, the subtitles should now always render above all player menu UI elements, but sadly I don't have a simple way of verifying that on mobile devices, if u don't mind checking if this solution is enough I would be really glad. 

Thanks for ur feedback!
